packages/create-subscription/src/createSubscription.js

```diff
@@ -0,0 +1,153 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import React from 'react';
+import invariant from 'fbjs/lib/invariant';
+import warning from 'fbjs/lib/warning';
+
+type Unsubscribe = () => void;
+type CannotUnsubscribe = false;
+
+export function createSubscription<Property, Value>(
+  config: $ReadOnly<{|
+    // Synchronously gets the value for the subscribed property.
+    // Return undefined if the subscribable value is undefined,
+    // Or does not support synchronous reading (e.g. native Promise).
+    getValue: (source: Property) => Value | void,
+
+    // Setup a subscription for the subscribable value in props, and return an unsubscribe function.
+    // Return false to indicate the property cannot be unsubscribed from (e.g. native Promises).
+    // Due to the variety of change event types, subscribers should provide their own handlers.
+    // Those handlers should not attempt to update state though;
+    // They should call the callback() instead when a subscription changes.
+    subscribe: (
+      source: Property,
+      callback: (value: Value | void) => void,
+    ) => Unsubscribe | CannotUnsubscribe,
+  |}>,
+): React$ComponentType<{
+  children: (value: Value | void) => React$Node,
+  source: any,
+}> {
+  const {getValue, subscribe} = config;
+
+  warning(
+    typeof getValue === 'function',
+    'Subscription must specify a getValue function',
+  );
+  warning(
+    typeof subscribe === 'function',
+    'Subscription must specify a subscribe function',
+  );
+
+  type Props = {
+    children: (value: Value) => React$Element<any>,
+    source: any,
+  };
+  type State = {
+    source: Property,
+    unsubscribeContainer: {
```

Comment:
Let's use `createRef`? :)

Reply:
What's the benefit of doing this? Seems like it would be more awkward, since I would need to create new refs each time from `getDerivedStateFromProps`.

Reply:
AFAICT it's the same pattern you're using. Roughly:

```js
state = {
  unsubscribe: React.createRef(null);
};

subscribe(source) {
  this.state.unsubscribe.current = unsubscribe;
}

componentWillUnmount() {
  this.state.unsubscribe.current();
}
```

Reply:
You're forgetting `getDerivedStateFromProps`

Reply:
Is it that you have to create a new object every time the source changes? That's what you're doing anyway

Reply:
Yeah, I just don't see the value of using `createRef` in this case, since we're throwing the ref away each time props change.

It also kind of blocks this PR on the `.value` to `.current` change 😄

Reply:
> Yeah, I just don't see the value of using createRef in this case, since we're throwing the ref away each time props change.

This is one of the points of `createRef`, to make the container pattern easier. If you don't see any benefit, maybe that means we shouldn't have a `createRef` API at all :D

> It also kind of blocks this PR on the .value to .current change

Not necessarily, I mean `createRef` is already in master. Would just have to update this when landing the naming change.

Reply:
I don't think `createRef` is inherently un-useful, I just didn't think it made _this particular case_ any better.

I also think it makes the Flow typing a little trickier.

Reply:
This is beyond the scope of this PR, but if it makes Flow typing harder... we should fix that. If `createRef` doesn't make this particular use case better, I question why we added it at all. Because this is the exact use case it's meant to be used for.

Reply:
> If `createRef` doesn't make this particular use case better, I question why we added it at all. Because this is the exact use case it's meant to be used for.

I disagree, but maybe in a subtle way.

I think `createRef` is nice when you're only the _consumer_ of the mutated value- because it requires a little less code. In this case, as both the consumer and the _manager_ of the value, I just didn't think it added anything.